### PR TITLE
Horizon Publisher

### PR DESCRIPTION
**Pull request by MrSimonEmms**
## Ticket Number

UCD-831

## Description of change

Submit completed appeal to Horizon

## Checklist

- [x] Requires infrastructure changes
- [x] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [x] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `master` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

**Review by frank-dodion — DISMISSED** [review]

File: Makefile [hunk 1]
```diff
@@ -50,3 +64,10 @@ uninstall:
 		(cd $${dir} && rm -Rf node_modules); \
   	done
 .PHONY: uninstall
+
+update-functions:
+	faas-cli remove -f functions.yml
+	faas-cli up -f functions.yml --tag sha \
```
Comment (a6software) [inline Makefile:70]: nice, I didn't know you could do multiple commands in a make file step with using && and chaining.

Comment (MrSimonEmms) [reply Makefile:70]: Yeah. It's similar (although not identical) to bash - the for loop above in that file highlights one of the main differences between Make and bash

File: charts/app/Chart.yaml [hunk 2]
```diff
@@ -4,3 +4,8 @@ description: PINS Appeal Process
 type: application
 version: 0.1.0
 icon: https://github.com/planning-inspectorate.png
+dependencies:
+  - name: openfaas-amqp1-0-connector
+    alias: horizon-householder-appeal-publish
+    version: ^1.0.0
+    repository: https://helm.simonemms.com
```
Comment (MrSimonEmms) [inline charts/app/Chart.yaml:11]: This is being put as my Helm repo for the time being. We have a task for PINS to fork this repo so it's not an external dependency

**Review by danWatkinson — COMMENTED** [review]
Have reviewed plumbing around our publishing of the new tests + the e2e tests brought in as part of this work. all looking good. That does comprise 12/128 files viewed according to github so not 'approving' on that basis, but it has my blessing when the time comes.

**Comment by frank-dodion** [comment]
I was able to trigger repeated messages on the queue for the same Appeal by using the PATCH endpoint to change the Appeal state from SUBMITTED to DRAFT and back again. Are we happy with this behaviour?

**Review by Fenisay — DISMISSED** [review]
LGTM

**Review by rcmsoftwaredevelopment — DISMISSED** [review]

**Comment by Fenisay** [comment]
> I was able to trigger repeated messages on the queue for the same Appeal by using the PATCH endpoint to change the Appeal state from SUBMITTED to DRAFT and back again. Are we happy with this behaviour?

My understanding is that we need to be able to modify the appeal after the submission somehow. The PATCH is a technical endpoint and not used by the forms web app. Maybe the solution should be to have a specific submit endpoint and push the queue only in this part and not mix all the needs in the same flow.  I really struggled to clean the controller to be more flexible now that we might have other interactions with horizon.

**Review by a6software — DISMISSED** [review]
If you're happy with everything then I am too :+1: mega.

**Review by MrSimonEmms — COMMENTED** [review]
ok

File: charts/app/Chart.lock [hunk 3]
```diff
@@ -0,0 +1,6 @@
+dependencies:
+- name: openfaas-amqp1-0-connector
+  repository: https://helm.simonemms.com
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply charts/app/Chart.lock:3]: Yeah. PINS will fork the repo, but this is a short-term solution

File: packages/horizon-add-document/handler.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            },
+          },
+        ],
+      },
+      'a:NodeId': '0',
+    },
+  };
+}
+
+debug({ config });
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply packages/horizon-add-document/handler.js:74]: It's not, but there's no sensitive data in this config. Any sensitive data is stored in a secret which isn't in the config (in OpenFaaS, secrets are put at `/var/openfaas/secrets/<secretName>`

File: infrastructure/environments/variables.tf [hunk 5]
```diff
@@ -112,6 +112,27 @@ variable "k8s_vm_size" {
...
+}
+
+variable "message_queue_capacity" {
+  description = "Message queue capacity - SKU must be premium if non-0"
+  type = number
+  default = 0
+}
+
+variable "message_queue_zone_redundancy_enabled" {
+  description = "Enable message queue redundancy - SKY must be premium if true"
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply infrastructure/environments/variables.tf:131]: Well spotted - should be SKU. Not the end of the world as in a description, but will change. Thanks

File: e2e-tests/cypress/integration/appeal-submission-to-horizon.feature [hunk 6]
```diff
@@ -0,0 +1,12 @@
+@wip
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply e2e-tests/cypress/integration/appeal-submission-to-horizon.feature:1]: @danWatkinson one for you to answer

File: packages/horizon-create-contact/handler.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
+const axios = require('axios');
+
+module.exports = async (event, context) => {
+  const config = {
+    horizon: {
+      url: process.env.HORIZON_URL,
+    },
+  };
+
+  console.log(JSON.stringify(config));
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply packages/horizon-create-contact/handler.js:10]: Annoyingly, in OpenFaas the logging is prefixed with the datetime stamp so we can't (easily) use Pino. It's easier using `console.log()` or `debug()` - the create-contact one is lightweight, so `console.log()` is fine

File: packages/horizon-create-contact/handler.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+  const { body } = event;
+
+  /* The order of this appears to be important */
+  // @todo add test for this
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply packages/horizon-create-contact/handler.js:15]: Nope - well-spotted. It's been done

File: packages/horizon-householder-appeal-publish/handler.js [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+
+      We treat these as non-mandatory for add documents, even though
+      they are mandatory in the appeal. This is to avoid any unhelpful
+      errors at this point
+    */
+    const documents = [
+      body?.appeal?.yourAppealSection?.appealStatement?.uploadedFile?.id,
+      body?.appeal?.requiredDocumentsSection?.originalApplication?.uploadedFile?.id,
+      body?.appeal?.requiredDocumentsSection?.decisionLetter?.uploadedFile?.id,
+    ];
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply packages/horizon-householder-appeal-publish/handler.js:358]: That's done as the first part of the `publishDocuments` function on line 218

File: packages/horizon-create-contact/handler.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      ),
+    },
+  };
+
+  console.log('Adding contact in Horizon', JSON.stringify(input));
+
+  try {
+    const { data } = await axios.post(`${config.horizon.url}/contacts`, input);
+
+    console.log('Contact added to Horizon', JSON.stringify(data));
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply packages/horizon-create-contact/handler.js:55]: I don't think there's a GDPR issue here. I agree it's not ideal, but I sort of want to see what's going on in here until we have a better clue of how Horizon works. I think this is justifiable as there's no password being stored in the logs, just name/email (it does make me wince saying it, but also, I have to manage this call)

File: e2e-tests/cypress/integration/appeal-submission-to-horizon.feature (hunk 6 above)
[the remark this answers is not in the document]
Comment (danWatkinson) [reply e2e-tests/cypress/integration/appeal-submission-to-horizon.feature:1]: hmm. yes. 
good idea.

[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply e2e-tests/cypress/integration/appeal-submission-to-horizon.feature:1]: We've put this back in as the config for the test has been lost. @danWatkinson will update the team

**Review by a6software — APPROVED** [review]

**Review by rcmsoftwaredevelopment — APPROVED** [review]

**Review by a6software — APPROVED** [review]

**Comment by MrSimonEmms** [comment]
Here goes nothing

**Pull request merged**